common/log.h

```diff
@@ -47,6 +50,332 @@
+                        {
+                            auto result = std::fopen(_filename.c_str(), "w");
+                            if( result )
+                            {
+                                _handle = result;
+                            }
+                            else
+                            {
+                                std::fprintf(stderr, "Failed to open logfile '%s' with error '%s'\n", _filename.c_str(), std::strerror(errno));
+                                std::fflush(stderr);
+                                _handle = stderr;
+                            }
+                            _opened = true;
+                        }
+                        _lock.clear(std::memory_order_release);
+                    }
+                    return _handle;
+                }
+
+                void flush()
+                {
+                    while( _lock.test_and_set(std::memory_order_acquire) ){ std::this_thread::yield(); }
+                    if(_opened && _type != Type::Invalid && _handle)
+                    {
+                        std::fflush(_handle);
+                    }
+                    _lock.clear(std::memory_order_release);
+                }
+
+            private:
+                std::atomic_flag   _lock      = ATOMIC_FLAG_INIT;
+                Type               _type      {Type::Invalid};
+                std::string        _filename;
+                std::atomic<bool>  _opened    {false};
+                std::atomic<FILE*> _handle    {nullptr};
+
+            //friend class LogStateWrapper;
+        };
+
+    private:
+        std::mutex                     _mutex;
+        std::atomic_flag               _lock                    = ATOMIC_FLAG_INIT;
+        std::atomic<bool>              _enabled                 {true};
+        std::atomic<LogTriState>       _global_override_enabled {LogTriState::Same};
+        std::set<LogTargetWrapper*>    _targets;
+        LogTargetWrapper               _null_target             {nullptr};
+        LogTargetWrapper               _stderr_target           {stderr};
+        std::atomic<LogTriState>       _global_override_target  {LogTriState::Same};
+        std::atomic<LogTargetWrapper*> _current_target          {&_null_target};
+        std::atomic<LogTargetWrapper*> _stored_target           {&_null_target};
+
+        LogTargetWrapper* log_set_target_impl( const std::string && filename ) {return log_set_target_impl(filename);}
+        LogTargetWrapper* log_set_target_impl( const std::string & filename )  {return log_add_select_target(new LogTargetWrapper(filename), true);}
+        LogTargetWrapper* log_set_target_impl( FILE* handle )                  {return log_add_select_target(new LogTargetWrapper(handle), true);}
+        LogTargetWrapper* log_set_target_impl( LogTargetWrapper * target )     {return log_add_select_target(target);}
```

Comment:
There should be spaces inside of the curly brackets, but not inside of the round brackets.